src/lem/eval.rs

```diff
@@ -828,6 +807,46 @@ fn reduce(cprocs: &[(&Symbol, usize)]) -> Func {
         return (nil)
     });
     let is_cproc = is_cproc(cprocs);
+    let lookup = func!(lookup(expr, env, found, binding): 4 => {
+        let rec = Symbol("rec");
+        let non_rec = Symbol("non_rec");
+        let error = Symbol("error");
+        let not_found = Symbol("not_found");
+        let continue = eq_val(not_found, found);
+        if !continue {
+            return (expr, env, found, binding)
+        }
+        match env.tag {
+            Expr::Nil => {
+                return (expr, env, error, binding)
+            }
+        };
+        let (binding, smaller_env) = car_cdr(env);
+        match binding.tag {
+            Expr::Nil => {
+                return (expr, env, error, binding)
+            }
+        };
+        let (var, val) = decons2(binding);
+        match var.tag {
+            Expr::Sym => {
+                let eq_val = eq_val(var, expr);
+                if eq_val {
+                    return (val, env, non_rec, binding)
+                }
+                return (expr, smaller_env, not_found, binding)
+            }
+            // This is a hack to signal that the binding came from a letrec
+            Expr::Thunk => {
```

Comment:
1. How much more expensive would it be to add a proper case in the environment?
2. If the answer is "quite a bit", could we document this special meaning in the definition of the `Thunk` variant in the enum declaration?

I also wouldn't mind removing the mention of "hack" here, and renaming the "Thunk" variant to `ThunkOrRecSym` (or something more evocative).

Reply:
@huitseeker not expensive at all, I think it would just add 1 constraint and 1 auxiliary. I just didn't do it because I'd have to add a bunch of cases in a bunch of files. I think I will add a new tag "FrozenVar", since it is a frozen computation that the variable refers to. I'll replace the comment as well